### PR TITLE
Create DateParser and TimeParser in ixdtf

### PR DESCRIPTION
**Pull request by samchen61661**
Create DateParser, TimeParser and does not use iterator in ixdtf

### CODE REVIEW

File: experimental/ixdtf/src/parser.rs [hunk 1]
```diff
@@ -50,392 +53,532 @@ impl DecimalSeparator {
...
+        while cnt < s.len() {
+            if let Some((first, remains)) = time_slice.split_first() {
+                time_slice = remains;
+                if Self::is_date_time_separator(first) {
+                    break;
                 }
             }
+            cnt += 1;
+        }
+        result.parsed_date = match DateParser::parse(&s[..cnt]) {
```
Comment (samchen61661) [inline experimental/ixdtf/src/parser.rs:573]: Don't know how to get the date slice without using `get()` or indexing

File: experimental/ixdtf/src/parser.rs [hunk 2]
```diff
@@ -50,392 +53,532 @@ impl DecimalSeparator {
...
-    where
-        I: Iterator<Item = &'a u8>,
-    {
+    fn parse_date_month(
+        s: &[u8],
+        start_index: usize,
+    ) -> Result<(Option<SegmentIndex>, &[u8]), ParseError> {
         let mut month: u8 = 0;
         let mut cnt = 0;
+        let mut mut_inner_remains = s;
```
Comment (Manishearth) [inline experimental/ixdtf/src/parser.rs:167]:
```suggestion
        let mut mut_inner_remains = self.bytes;
```

File: experimental/ixdtf/src/parser.rs [hunk 3]
```diff
@@ -50,392 +53,532 @@ impl DecimalSeparator {
...
                 return Err(ParseError::DateMonth);
             }
-            iter.next();
             cnt += 1;
         }
         if !(1..=12).contains(&month) {
             return Err(ParseError::DateMonth);
         }
-        return Ok(Some(month));
+        return Ok((
```
Comment (Manishearth) [inline experimental/ixdtf/src/parser.rs:184]:
```
self.bytes = mut_inner_remains;

return Ok(month)
```

File: experimental/ixdtf/src/parser.rs [hunk 4]
```diff
@@ -50,392 +53,532 @@ impl DecimalSeparator {
...
     }
 
-    fn parse_date_month<'a, I>(iter: &mut core::iter::Peekable<I>) -> Result<Option<u8>, ParseError>
-    where
-        I: Iterator<Item = &'a u8>,
-    {
+    fn parse_date_month(
+        s: &[u8],
+        start_index: usize,
+    ) -> Result<(Option<SegmentIndex>, &[u8]), ParseError> {
```
Comment (Manishearth) [inline experimental/ixdtf/src/parser.rs:164]:
```suggestion
    ) -> Result<u8, ParseError> {
```

**Review by Manishearth — COMMENTED** [review]
demo of changes

File: experimental/ixdtf/src/parser.rs (hunk 4 above)
Comment (samchen61661) [reply experimental/ixdtf/src/parser.rs]: I prefer `Option<u8>` here because time could be optional.

File: experimental/ixdtf/src/parser.rs [hunk 5]
```diff
@@ -61,314 +64,241 @@ struct ParsedDateTime {
...
-        while cnt < 6 {
-            if let Some(y) = iter.peek() {
-                if **y >= b'0' && **y <= b'9' {
-                    year = year * 10 + **y as i32 - b'0' as i32;
+struct DateTimeParser<'a> {
+    bytes: &'a [u8],
+}
+
+impl<'a> DateTimeParser<'a> {
+    fn parse_date_extended_year(&mut self) -> Result<Option<i32>, ParseError> {
```
Comment (Manishearth) [inline experimental/ixdtf/src/parser.rs:72]: nit: in the long run all of these should be documented, saying what they're parsing with examples (and explaining things like why there's an Option here)

for now it's fine to punt docs till later

File: experimental/ixdtf/src/parser.rs [hunk 6]
```diff
@@ -556,50 +543,54 @@ mod test {
     #[test]
     fn test_bad_date() {
         let dt = "-2022-06-05".as_bytes();
-        let parsed = DateTimeParser::parse(dt);
+        let parsed = DateTimeParser { bytes: dt }.parse();
```
Comment (Manishearth) [inline experimental/ixdtf/src/parser.rs:546]: nit: we should probably have a constructor like `DateTimeParser::new()`

**Review by Manishearth — COMMENTED** [review]
This is the right architecture!

@nordzilla do you want to review the actual idxtf parsing algorithm?

**Review by sffc — COMMENTED** [review]
Agree: Architecturally, this is the best version so far

File: experimental/ixdtf/src/parser.rs (hunk 5 above)
Comment (samchen61661) [reply experimental/ixdtf/src/parser.rs:94]: add simple comments for the structs in this file. Will add more once the code is stable

File: experimental/ixdtf/src/parser.rs (hunk 6 above)
Comment (samchen61661) [reply experimental/ixdtf/src/parser.rs]: sg!

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
kicking CI

**Pull request merged**